### ISSUE TITLE
Notifications are not sent

### ISSUE THREAD

**Issue by apollo13** [ansible/awx#11422]
### Please confirm the following

- [X] I agree to follow this project's [code of conduct](http://docs.ansible.com/ansible/latest/community/code_of_conduct.html).
- [X] I have checked the [current issues](https://github.com/ansible/awx/issues) for duplicates.
- [X] I understand that AWX is open source software provided for free and that I am not entitled to status updates or other assurances.

### Summary

Currently notifications for (some) jobs are not sent due to a race condition (?):
```
awx_1    | 2021-12-06 13:39:14,446 WARNING  [a1a65966262c4559afd398df32b5f445] awx.main.dispatch scaling down worker pid:14114
awx_1    | 2021-12-06 13:39:14,448 DEBUG    [a1a65966262c4559afd398df32b5f445] awx.main.dispatch task 4648c5ba-e5bb-49c4-8040-0e98e1d4975b starting awx.main.tasks.handle_success_and_failure_notifications(*[4021])
awx_1    | 2021-12-06 13:39:14,450 WARNING  [a1a65966262c4559afd398df32b5f445] awx.main.dispatch scaling down worker pid:14116
awx_1    | 2021-12-06 13:39:14,451 WARNING  [a1a65966262c4559afd398df32b5f445] awx.main.dispatch worker exiting gracefully pid:14114
awx_1    | 2021-12-06 13:39:14,453 WARNING  [a1a65966262c4559afd398df32b5f445] awx.main.dispatch worker exiting gracefully pid:14116
awx_1    | 2021-12-06 13:39:14,813 INFO     [949fbc302a95418e85b3b941b7bb551d] awx.main.commands.run_callback_receiver Event processing is finished for Job 4021, sending notifications
awx_1    | 2021-12-06 13:39:14,813 INFO     [949fbc302a95418e85b3b941b7bb551d] awx.main.commands.run_callback_receiver Event processing is finished for Job 4021, sending notifications
awx_1    | INFO 2021/12/06 13:39:15 Stdout complete - closing channel for: aCroB8Bf 
awx_1    | ERROR 2021/12/06 13:39:15 Read error in control service: read unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
awx_1    | INFO 2021/12/06 13:39:15 Client disconnected from control service @
awx_1    | ERROR 2021/12/06 13:39:15 Error closing connection: close unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
awx_1    | 2021-12-06 13:39:15,939 DEBUG    [-] awx.main.commands.run_callback_receiver JobEvent.objects.bulk_create(1)
awx_1    | 2021-12-06 13:39:15,939 DEBUG    [-] awx.main.commands.run_callback_receiver JobEvent.objects.bulk_create(1)
awx_1    | 2021-12-06 13:39:18,015 DEBUG    [4f92e389a2534afd82ac34149033c2f3] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/4021/job_events/?counter__gte=13&limit=9&order_by=counter'>, response_time: 0.218s
awx_1    | [pid: 45|app: 0|req: 300/912] 192.168.207.101 () {56 vars in 2559 bytes} [Mon Dec  6 13:39:17 2021] GET /api/v2/jobs/4021/job_events/?counter__gte=13&limit=9&order_by=counter => generated 14232 bytes in 220 msecs (HTTP/2.0 200) 15 headers in 581 bytes (1 switches on core 0)
awx_1    | INFO 2021/12/06 13:39:18 Client connected to control service @
awx_1    | INFO 2021/12/06 13:39:18 Control service closed
awx_1    | INFO 2021/12/06 13:39:18 Client disconnected from control service @
awx_1    | 2021-12-06 13:39:18,481 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.tasks job 4021 (running) finished running, producing 22 events.
awx_1    | 2021-12-06 13:39:18,494 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.analytics.job_lifecycle job-4021 post run
awx_1    | 2021-12-06 13:39:18,676 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.analytics.job_lifecycle job-4021 finalize run
awx_1    | 2021-12-06 13:39:18,776 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.dispatch task e9c3cb47-b314-4f01-9a80-67388d7c42e0 starting awx.main.tasks.update_inventory_computed_fields(*[2])
awx_1    | 2021-12-06 13:39:18,806 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.models.inventory Going to update inventory computed fields, pk=2
awx_1    | 2021-12-06 13:39:18,839 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.models.inventory Finished updating inventory computed fields, pk=2, in 0.032 seconds
awx_1    | 2021-12-06 13:39:18,931 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.dispatch task 7fdefc75-7d90-41f1-8bd5-11f173086630 starting awx.main.tasks.handle_work_success(*[])
awx_1    | 2021-12-06 13:39:18,951 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.dispatch task 8246994d-45e1-4576-be9a-e53a623f7015 starting awx.main.scheduler.tasks.run_task_manager(*[])
awx_1    | 2021-12-06 13:39:18,955 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.scheduler Running task manager.
awx_1    | 2021-12-06 13:39:18,966 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.scheduler Starting Scheduler
awx_1    | 2021-12-06 13:39:19,034 DEBUG    [fc7226b03a7247d0bf5f473108e35e76] awx.analytics.job_lifecycle job-4021 event processing finished
awx_1    | 2021-12-06 13:39:19,042 DEBUG    [949fbc302a95418e85b3b941b7bb551d] awx.main.scheduler Finishing Scheduler
awx_1    | 2021-12-06 13:39:19,070 DEBUG    [fc7226b03a7247d0bf5f473108e35e76] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/4021/'>, response_time: 0.207s
awx_1    | [pid: 45|app: 0|req: 301/913] 192.168.207.101 () {56 vars in 2456 bytes} [Mon Dec  6 13:39:18 2021] GET /api/v2/jobs/4021/ => generated 4964 bytes in 210 msecs (HTTP/2.0 200) 14 headers in 565 bytes (1 switches on core 0)
awx_1    | 2021-12-06 13:39:19,561 WARNING  [a1a65966262c4559afd398df32b5f445] awx.main.tasks Failed to even try to send notifications for job '2021-12-06 13:38:04.895362+00:00-4021-successful' due to job not being in finished state.
```

Are the receptor error messages in there expected?

### AWX version

19.5.0

### Installation method

docker on linux

### Modifications

no

### Ansible version

4.9.0

### Operating system

CentOS 7

### Web browser

Firefox, Chrome, Safari, Edge

### Steps to reproduce

Create a notification and change a job to send notifications on success/failure

### Expected results

Notifications are sent out

### Actual results

Notifications are not sent out for some jobs

**Comment by apollo13** [comment]
I just managed to reproduce this on devel with the standard development setup, every now and then (again not for all jobs), the output looks like this (test env is started via `CONTROL_PLANE_NODE_COUNT=1 EXECUTION_NODE_COUNT=0 COMPOSE_TAG=devel make docker-compose`):
```
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,078 DEBUG    [162c0b54] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/24/relaunch/'>, response_time: 0.096s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/24/relaunch/ - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,240 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 created
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,310 DEBUG    [524c6c4a] awx.analytics.performance request: <WSGIRequest: POST '/api/v2/jobs/24/relaunch/'>, response_time: 0.226s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 POST /api/v2/jobs/24/relaunch/ - HTTP/1.1 201
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,365 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 acknowledged
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,402 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 created
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,441 DEBUG    [d2610f29] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/unified_jobs/?id=28'>, response_time: 0.115s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/unified_jobs/?id=28 - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,474 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 execution node chosen
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,475 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 controller node chosen
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,511 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 waiting
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,515 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 blocked by projectupdate-29
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,545 DEBUG    [558480b2] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/unified_jobs/?id=28'>, response_time: 0.099s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/unified_jobs/?id=28 - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,671 DEBUG    [65905340] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/28/'>, response_time: 0.111s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/28/ - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,716 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 pre run
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,721 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 preparing playbook
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,747 DEBUG    [da3cbef8] awx.analytics.performance request: <WSGIRequest: OPTIONS '/api/v2/jobs/28/job_events/'>, response_time: 0.070s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 OPTIONS /api/v2/jobs/28/job_events/ - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,752 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 running playbook
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:31:58 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,850 DEBUG    [22d83e0f] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/28/job_events/?limit=1&order_by=-counter'>, response_time: 0.073s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/28/job_events/?limit=1&order_by=-counter - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:31:58 Control service closed
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:31:58 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:31:58,860 DEBUG    [7c099be5] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/28/job_events/?counter__gte=1&limit=50&order_by=counter'>, response_time: 0.086s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,860 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 work unit id received
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/28/job_events/?counter__gte=1&limit=50&order_by=counter - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:58,882 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 work unit id assigned
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:31:58 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:31:59,250 DEBUG    [e96c8f9f] awx.analytics.job_lifecycle job-28 blocked by projectupdate-29
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:02 Stdout complete - closing channel for: LNEwf2YP 
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receptor stdout | ERROR 2021/12/06 18:32:02 Read error in control service: read unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:02 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | ERROR 2021/12/06 18:32:02 Error closing connection: close unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receiver stdout | 2021-12-06 18:32:02,652 INFO     [524c6c4a] awx.main.commands.run_callback_receiver Event processing is finished for Job 29, sending notifications
tools_awx_1 | awx-receiver stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:05 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:05 Control service closed
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:05 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:05,934 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 post run
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,016 DEBUG    [524c6c4a] awx.analytics.job_lifecycle projectupdate-29 finalize run
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,107 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 execution node chosen
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,107 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 controller node chosen
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,135 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 waiting
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,292 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 pre run
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,426 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 preparing playbook
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,524 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 running playbook
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:06 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:06 Control service closed
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:06 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,580 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 work unit id received
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:06,623 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 work unit id assigned
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:06 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:13 Stdout complete - closing channel for: BhNVTMh8 
tools_awx_1 | awx-receptor stdout | ERROR 2021/12/06 18:32:13 Read error in control service: read unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:13 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | ERROR 2021/12/06 18:32:13 Error closing connection: close unix /var/run/awx-receptor/receptor.sock->@: use of closed network connection
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receiver stdout | 2021-12-06 18:32:13,897 INFO     [524c6c4a] awx.main.commands.run_callback_receiver Event processing is finished for Job 28, sending notifications
tools_awx_1 | awx-receiver stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:32:13,907 DEBUG    [09ea85f4] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/28/job_events/?counter__gte=0&limit=9&order_by=counter'>, response_time: 0.118s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/28/job_events/?counter__gte=0&limit=9&order_by=counter - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 

tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:17 Client connected to control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:17 Control service closed
tools_awx_1 | awx-receptor stdout | INFO 2021/12/06 18:32:17 Client disconnected from control service @
tools_awx_1 | awx-receptor stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:17,711 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 post run
tools_awx_1 | awx-dispatcher stdout | 

tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:17,814 DEBUG    [524c6c4a] awx.analytics.job_lifecycle job-28 finalize run
tools_awx_1 | awx-dispatcher stdout | 

tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:32:17,967 DEBUG    [bfb66a2c] awx.analytics.job_lifecycle job-28 event processing finished
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:32:17,982 DEBUG    [bfb66a2c] awx.analytics.performance request: <WSGIRequest: GET '/api/v2/jobs/28/'>, response_time: 0.101s
tools_awx_1 | awx-uwsgi stdout | 
tools_awx_1 | awx-uwsgi stdout | 172.27.0.1 GET /api/v2/jobs/28/ - HTTP/1.1 200
tools_awx_1 | awx-uwsgi stdout | 

tools_awx_1 | awx-dispatcher stdout | 2021-12-06 18:32:18,799 WARNING  [524c6c4a] awx.main.tasks Failed to even try to send notifications for job '2021-12-06 18:31:58.148550+00:00-28-successful' due to job not being in finished state.
tools_awx_1 | awx-dispatcher stdout | 
```

**Comment by apollo13** [comment]
I found this gem in the code:
https://github.com/ansible/awx/blob/19b4849345ecc7f9b2bd6b8df77e31116d494f62/awx/main/tasks.py#L794-L810

So it seems the race condition is known/expected but the 5 seconds don't seem to cut it. My dev environment is an idle laptop with 16 GB RAM and an Intel(R) Core(TM) i7-8565U CPU. I do not think that system should be to slow -- rather there is probably a systematic underlying problem.

**Comment by lyutian** [comment]
According to the log, the notification began to send at `2021-12-06 18:32:13,897`:
```
tools_awx_1 | awx-receiver stdout | 2021-12-06 18:32:13,897 INFO     [524c6c4a] awx.main.commands.run_callback_receiver Event processing is finished for Job 28, sending notifications
```

But, the job finished at `2021-12-06 18:32:17,967`, which is more than 4 seconds(**the timeout is 4 sec, not 5, if you read the code**).
```
tools_awx_1 | awx-uwsgi stdout | 2021-12-06 18:32:17,967 DEBUG    [bfb66a2c] awx.analytics.job_lifecycle job-28 event processing finished
```

So , job completed notification sent failed.

---

For the reason why spent so much time, it's not only related to the host performance, but also the communication between receptor and awx. The 4 seconds is not a reliable timeout setting, I guess the developor want 5, but the code doesn't.
```
                                            ┌──────────────┐
                                            │              │
┌──────────────┐                 ┌──────────┤  receptor-1  │
│              │                 │          │              │
│    awx_1     │◄──────────┐     │          └──────────────┘
│              │           │     ▼
└──────┬───────┘    ┌──────┴───────┐        ┌──────────────┐
       │            │              │        │              │
       │            │ receptor-hop │◄───────┤  receptor-2  │
       ▼            │              │        │              │
┌──────────────┐    └──────────────┘        └──────────────┘
│              │                 ▲
│    awx_2     │                 │          ┌──────────────┐
│              │                 │          │              │
└──────────────┘                 └──────────┤  receptor-3  │
                                            │              │
                                            └──────────────┘
```

**Comment by apollo13** [comment]
@lyutian I agree that is related to more than just host-related. In my case it is the minimal docker setup with a single hybrid control node and no executors; so this is all in the same docker container -- no hops etc…

**Comment by lyutian** [comment]
@apollo13 Get into your awx container `docker exec -it tools_awx_1 bash`, modify the `/awx_devel/awx/main/tasks.py` by change the timeout from 5 to bigger, can temporarily resolve this issue.

**Comment by apollo13** [comment]
Thanks, I know that I can do that as workaround; but as you say it is just a workaround.

**Comment by azdio-support** [comment]
@lyutian I am using AWX Operator 19.5 and have the same issue. Do you know how to edit tasks.py inside the container? The file is read-only for awx user and I don't know the password.

**Comment by lyutian** [comment]
@azdio-support If you use k8s operator to deploy the awx in production mode, only `awx` user you have in the awx container(check `/etc/passwd`, you will find). The operator production deploy is not changeable, you need to use a devel image, please refer: https://github.com/ansible/awx/blob/devel/docs/development/minikube.md#custom-awx-development-image-for-kubernetes

**Comment by azdio-support** [comment]
@lyutian Thanks for the information, I was missed that. 
I believe that this will be fixed in some of the next releases, so I can wait for it.

**Comment by coudenysj** [comment]
Are you sure this will be resolved in a future version?

I'm also seeing this error in our setup:

```
2022-02-16 02:40:16,234 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 created
2022-02-16 02:40:16,234 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.tasks Spawned job 56095 (new) from schedule 171.
2022-02-16 02:40:16,394 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 acknowledged
2022-02-16 02:40:16,446 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 waiting
2022-02-16 02:40:16,447 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.scheduler job 56095 (waiting) consumed 31 capacity units from x with prior total of 31
2022-02-16 02:40:16,533 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.dispatch task 6c72e292-657f-412a-9d25-169567938247 starting awx.main.tasks.RunJob(*[56095])
2022-02-16 02:40:17,256 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 pre run
2022-02-16 02:40:17,270 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.tasks Skipping project sync for job 56095 (running) because commit is locally available
2022-02-16 02:40:17,779 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.notifications.slack_backend {payload}
2022-02-16 02:40:17,790 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 notifications sent
2022-02-16 02:40:18,284 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.tasks Project 479 prepared /tmp/pdd_wrapper_56095_pnk4lmzt/awx_56095_slbry5g2/requirements_roles from cache
2022-02-16 02:40:18,286 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 preparing playbook
2022-02-16 02:40:18,287 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 start job fact cache
2022-02-16 02:40:47,615 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 running playbook
2022-02-16 03:03:20,511 DEBUG    [48f333404b7c495eb4a1839b1d2a3cce] awx.main.dispatch task f2dd1a9f-7302-4dfc-90f9-ef1052369c1c starting awx.main.tasks.handle_success_and_failure_notifications(*[56095])
2022-02-16 03:03:25,577 WARNING  [48f333404b7c495eb4a1839b1d2a3cce] awx.main.tasks Failed to even try to send notifications for job '2022-02-16 02:40:16.081250+00:00-56095-running' due to job not being in finished state.
2022-02-16 03:03:26,689 INFO     [07479b253ab9461f8995e77fc9eb7cbf] awx.main.commands.run_callback_receiver Event processing is finished for Job 56095, sending notifications
2022-02-16 03:03:26,689 INFO     [07479b253ab9461f8995e77fc9eb7cbf] awx.main.commands.run_callback_receiver Event processing is finished for Job 56095, sending notifications
2022-02-16 03:03:27,561 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.main.tasks job 56095 (running) finished running, producing 35424 events.
2022-02-16 03:03:27,567 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 post run
2022-02-16 03:03:27,654 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 finalize run
2022-02-16 03:03:27,655 DEBUG    [07479b253ab9461f8995e77fc9eb7cbf] awx.analytics.job_lifecycle job-56095 finish job fact cache
```

**Comment by coudenysj** [comment]
I think the "fix" in #11750 won't be enough, because today I had a job that took 20 seconds:

```
2022-02-17 06:54:49,575 DEBUG    [b8c34f2e75d2430c8874f3f8feb557cc] awx.main.dispatch task 61ea3840-8abf-4847-ad26-023ffb78cba7 starting awx.main.tasks.handle_success_and_failure_notifications(*[56302])
2022-02-17 06:54:54,645 WARNING  [b8c34f2e75d2430c8874f3f8feb557cc] awx.main.tasks Failed to even try to send notifications for job '2022-02-17 06:37:20.281267+00:00-56302-running' due to job not being in finished state.
2022-02-17 06:55:09,187 INFO     [15249c3627f04c718f0532cb3c889677] awx.main.commands.run_callback_receiver Event processing is finished for Job 56302, sending notifications
```

Should the calling of `handle_success_and_failure_notifications` not be postponed after the `Event processing is finished for Job x`?

**Comment by AlanCoding** [comment]
Link https://github.com/ansible/awx/pull/11933

**Comment by hesmithrh** [comment]
@AlanCoding to run regression tests and setup merge meeting to review with a group

**Comment by rebeccahhh** [comment]
@AlanCoding does #11933 resolve this fully? AKA should we close this?

**Comment by AlanCoding** [comment]
Yes, this is in Needs Test.

As a note for QE, in my own testing, I would just add a `time.sleep(6)` or something like this in the `awx.main.tasks.jobs` code path before the final status change. It is fairly easy to verify (using integration tests) that this causes notifications to not be sent before this change, and that the change fixes it. I'm sure there are other ways to force the bug to happen, but that could be hard.
